crates/matrix-sdk-sled/src/cryptostore.rs

```diff
@@ -480,96 +479,18 @@ impl SledStore {
-                deleted: bool,
-                trust_state: LocalTrust,
-            }
-
-            #[allow(clippy::from_over_into)]
-            impl Into<ReadOnlyDevice> for OldReadOnlyDevice {
-                fn into(self) -> ReadOnlyDevice {
-                    let mut device_keys = DeviceKeys::new(
-                        self.user_id,
-                        self.device_id,
-                        self.algorithms,
-                        self.keys,
-                        self.signatures,
-                    );
-                    device_keys.unsigned.device_display_name = self.display_name;
-
-                    ReadOnlyDevice::new(device_keys, self.trust_state)
-                }
-            }
-
-            let devices: Vec<ReadOnlyDevice> = self
-                .devices
-                .iter()
-                .map(|d| {
-                    serde_json::from_slice(&d.map_err(|e| CryptoStoreError::Backend(anyhow!(e)))?.1)
-                        .map_err(CryptoStoreError::Serialization)
-                })
-                .map(|d| {
-                    let d: OldReadOnlyDevice = d?;
-                    Ok(d.into())
-                })
-                .collect::<Result<Vec<ReadOnlyDevice>, CryptoStoreError>>()?;
-
-            self.devices
-                .transaction(move |tree| {
-                    for device in &devices {
-                        let key = device.encode();
-                        let device = serde_json::to_vec(device)
-                            .map_err(ConflictableTransactionError::Abort)?;
-                        tree.insert(key, device)?;
-                    }
-
-                    Ok(())
-                })
-                .map_err(|e| CryptoStoreError::Backend(anyhow!(e)))?;
-        }
-
-        if version <= 2 {
-            // We're treating our own device now differently, we're checking if
-            // the keys match to what we have locally, remove the unchecked
-            // device and mark our own user as dirty.
-            if let Some(pickle) = self
-                .account
-                .get("account".encode())
-                .map_err(|e| CryptoStoreError::Backend(anyhow!(e)))?
-            {
-                let pickle = serde_json::from_slice(&pickle)?;
-                let account = ReadOnlyAccount::from_pickle(pickle)?;
-
-                self.devices
-                    .remove((account.user_id().as_str(), account.device_id.as_str()).encode())
-                    .map_err(|e| CryptoStoreError::Backend(anyhow!(e)))?;
-                self.tracked_users
-                    .insert(account.user_id().as_str(), &[true as u8])
-                    .map_err(|e| CryptoStoreError::Backend(anyhow!(e)))?;
-            }
+        if version <= 3 {
+            return Err(anyhow!(
+                "Unsupported database version, the database \
+                 format changed in a backwards incompatible way"
+            )
```

Comment:
Should we maybe have an actual error-type for Upgrade-issues so that a higher level can react upon it by backing up and deleting the existing instance and restarting automatically? Right now, it could only show that message and can't assist the user in resolving the issue...

Reply:
Dunno, depends on the error (e.g. `UnsupportedDatabaseVersion(current, latest)`) and the way we want to handle this. I can still foresee a scenario where you might have really old databases and don't want to enforce upgrades (e.g. anything from before this release or because too many upgrades have been skipped) or you want the surrounding lib decide upon the details of the migration, might even interact with the user to provide a path for backups or something...

Reply:
Alright, will add.

Reply:
Did so in https://github.com/matrix-org/matrix-rust-sdk/pull/494/commits/c43b1b9e9d3104d504b3336e593c42ae6605c429.